### PR TITLE
V10: Find and persist embedded images in rich text

### PR DESCRIPTION
**Pull request by iOvergaard**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description

Due to troublesome CPU spikes when encountering base64 images, this PR brings a feature that will persist embedded images found in rich text as files in the Media library similar to what we are doing with drag & drop images.

### CODE REVIEW

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 1]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
 
         _tempFolderAbsolutePath = _hostingEnvironment.MapPathContentRoot(Constants.SystemDirectories.TempImageUploads);
 
+        _imageUrlGenerator = StaticServiceProvider.Instance.GetRequiredService<IImageUrlGenerator>();
+        _contentSettings = StaticServiceProvider.Instance.GetRequiredService<IOptions<ContentSettings>>().Value;
+    }
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:72]: This ctor should use `this` instead, to call the new constructor 🙌 

So it should look like this: 

```
    public RichTextEditorPastedImages(
        IUmbracoContextAccessor umbracoContextAccessor,
        ILogger<RichTextEditorPastedImages> logger,
        IHostingEnvironment hostingEnvironment,
        IMediaService mediaService,
        IContentTypeBaseServiceProvider contentTypeBaseServiceProvider,
        MediaFileManager mediaFileManager,
        MediaUrlGeneratorCollection mediaUrlGenerators,
        IShortStringHelper shortStringHelper,
        IPublishedUrlProvider publishedUrlProvider)
        : this(
            umbracoContextAccessor,
            logger,
            hostingEnvironment,
            mediaService,
            contentTypeBaseServiceProvider,
            mediaFileManager,
            mediaUrlGenerators,
            shortStringHelper,
            publishedUrlProvider,
            StaticServiceProvider.Instance.GetRequiredService<IImageUrlGenerator>(),
            StaticServiceProvider.Instance.GetRequiredService<IOptions<ContentSettings>>())
    {
    }
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 2]
```diff
@@ -30,7 +37,11 @@ public sealed class RichTextEditorPastedImages
     private readonly IShortStringHelper _shortStringHelper;
     private readonly IUmbracoContextAccessor _umbracoContextAccessor;
     private readonly string _tempFolderAbsolutePath;
+    private readonly IImageUrlGenerator _imageUrlGenerator;
+    private readonly ContentSettings _contentSettings;
+    private readonly Dictionary<string, GuidUdi> _uploadedImages = new();
 
+    [Obsolete("Use the ctor which takes an IImageUrlGenerator and IOptions<ContentSettings> instead")]
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:44]: Lets add the version it gets removed in too, just so we remember it 👍 
```suggestion
    [Obsolete("Use the ctor which takes an IImageUrlGenerator and IOptions<ContentSettings> instead, scheduled for removal in v14")]
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 3]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
     {
         // Find all img's that has data-tmpimg attribute
         // Use HTML Agility Pack - https://html-agility-pack.net
         var htmlDoc = new HtmlDocument();
         htmlDoc.LoadHtml(html);
 
-        HtmlNodeCollection? tmpImages = htmlDoc.DocumentNode.SelectNodes($"//img[@{TemporaryImageDataAttribute}]");
-        if (tmpImages == null || tmpImages.Count == 0)
+        HtmlNodeCollection? imagesWithDataUris = htmlDoc.DocumentNode.SelectNodes("//img");
+        if (imagesWithDataUris == null || imagesWithDataUris.Count == 0)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:120]: Lets change this to pattern matching while we're at it 🔥 
```suggestion
        if (imagesWithDataUris is null || imagesWithDataUris.Count is 0)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 4]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
 
-            var absoluteTempImagePath = Path.GetFullPath(_hostingEnvironment.MapPathContentRoot(tmpImgPath));
+            // Create tmp image by scanning the srcValue
+            // the value will look like "data:image/jpg;base64,abc" where the first part
+            // is the mimetype and the second (after the comma) is the image blob
+            Match dataUriInfo = Regex.Match(srcValue, @"^data:\w+\/(?<ext>\w+)[\w\+]*?;(?<encoding>\w+),(?<data>.+)$");
 
-            if (IsValidPath(absoluteTempImagePath) == false)
+            // If it turns up false, it was probably a false-positive and we can't do anything with it
+            if (dataUriInfo.Success == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:147]: Lets change this to pattern matching while we're at it 🔥
```suggestion
            if (dataUriInfo.Success is false)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 5]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
-
-            var mediaItemName = safeFileName.ToFriendlyName();
-            IMedia mediaFile;
-            GuidUdi udi;
+            var ext = dataUriInfo.Groups["ext"].Value.ToLowerInvariant();
+            var encoding = dataUriInfo.Groups["encoding"].Value.ToLowerInvariant();
+            var imageData = dataUriInfo.Groups["data"].Value;
 
-            if (uploadedImages.ContainsKey(tmpImgPath) == false)
+            if (_contentSettings.IsFileAllowedForUpload(ext) == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:156]: Lets change this to pattern matching while we're at it 🔥
```suggestion
            if (_contentSettings.IsFileAllowedForUpload(ext) is false)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 6]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+            // Create an unique folder path to help with concurrent users to avoid filename clash
+            var imageTempPath =
+                _hostingEnvironment.MapPathContentRoot(Constants.SystemDirectories.TempImageUploads + Path.DirectorySeparatorChar + Guid.NewGuid());
 
-                using (fileStream)
-                {
-                    mediaFile.SetValue(_mediaFileManager, _mediaUrlGenerators, _shortStringHelper, _contentTypeBaseServiceProvider, Constants.Conventions.Media.File, safeFileName, fileStream);
-                }
+            // Ensure image temp path exists
+            if (Directory.Exists(imageTempPath) == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:175]: Lets change this to pattern matching while we're at it 🔥
```suggestion
            if (Directory.Exists(imageTempPath) is false)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 7]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+    /// <returns>Formatted HTML.</returns>
+    public string FindAndPersistPastedTempImages(string html, Guid mediaParentFolder, int userId)
+    {
+        // Find all img's that has data-tmpimg attribute
+        // Use HTML Agility Pack - https://html-agility-pack.net
+        var htmlDoc = new HtmlDocument();
+        htmlDoc.LoadHtml(html);
+
+        HtmlNodeCollection? tmpImages = htmlDoc.DocumentNode.SelectNodes($"//img[@{TemporaryImageDataAttribute}]");
+        if (tmpImages == null || tmpImages.Count == 0)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:231]: Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (tmpImages is null || tmpImages.Count == 0)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 8]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+        }
+
+        return htmlDoc.DocumentNode.OuterHtml;
+    }
+
+    private void PersistMediaItem(Guid mediaParentFolder, int userId, HtmlNode img, string qualifiedTmpImgPath)
+    {
+        var absoluteTempImagePath = Path.GetFullPath(qualifiedTmpImgPath);
+
+        if (IsValidPath(absoluteTempImagePath) == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:258]: Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (IsValidPath(absoluteTempImagePath) is false)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 9]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+        var fileName = Path.GetFileName(absoluteTempImagePath);
+        var safeFileName = fileName.ToSafeFileName(_shortStringHelper);
 
-            // Find the width & height attributes as we need to set the imageprocessor QueryString
-            var width = img.GetAttributeValue("width", int.MinValue);
-            var height = img.GetAttributeValue("height", int.MinValue);
+        var mediaItemName = safeFileName.ToFriendlyName();
+        GuidUdi udi;
+
+        if (_uploadedImages.ContainsKey(qualifiedTmpImgPath) == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:269]: Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (_uploadedImages.ContainsKey(qualifiedTmpImgPath) is false)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 10]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+                mediaFile = _mediaService.CreateMedia(mediaItemName, mediaParentFolder, mediaType, userId);
             }
 
-            img.SetAttributeValue("src", location);
+            var fileInfo = new FileInfo(absoluteTempImagePath);
 
-            // Remove the data attribute (so we do not re-process this)
-            img.Attributes.Remove(TemporaryImageDataAttribute);
+            FileStream? fileStream = fileInfo.OpenReadWithRetry();
+            if (fileStream == null)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:289]: Lets change this to pattern matching while we're at it 🔥
```suggestion
            if (fileStream is null)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 11]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+            udi = _uploadedImages[qualifiedTmpImgPath];
+        }
+
+        // Add the UDI to the img element as new data attribute
+        img.SetAttributeValue("data-udi", udi.ToString());
+
+        // Get the new persisted image URL
+        _umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext? umbracoContext);
+        IPublishedContent? mediaTyped = umbracoContext?.Media?.GetById(udi.Guid);
+        if (mediaTyped == null)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:316]: Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (mediaTyped is null)
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs [hunk 12]
```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
...
+            });
+        }
+
+        img.SetAttributeValue("src", location);
+
+        // Remove the data attribute (so we do not re-process this)
+        img.Attributes.Remove(TemporaryImageDataAttribute);
+
+        // Add to the dictionary to avoid dupes
+        if (_uploadedImages.ContainsKey(qualifiedTmpImgPath) == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:344]: Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (_uploadedImages.ContainsKey(qualifiedTmpImgPath) is false)
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Found a few small semantic things, the rest looks good 😁

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs (hunk 3 above)
Comment (ronaldbarendse) [reply src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs:120]: You can even use [property pattern matching](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#property-pattern) to make this more concise 😄
```suggestion
        if (imagesWithDataUris is null or { Count: 0 })
```

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs (hunk 1 above)
Comment (iOvergaard) [reply src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs]: Fixed

File: src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs (hunk 2 above)
Comment (iOvergaard) [reply src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs]: Fixed

**Pull request merged**